mitodl/mit-open
# Add separate field for ocw topics, use best field to assign related topics

### What are the relevant tickets?
Closes https://github.com/mitodl/hq/issues/5499

### Description (What does it do?)
- For OCW, assigns topics from the course data's new `mit_learn_topics` field if populated, otherwise falls back on the usual `topics` field.  
- Adds a new `ocw_topics` field as an array of strings.
- Populates this new field from the OCW course data topics field, pretty much as is, just removing dupes and sorting.  This field will eventually be used by OCW search.


### How can this be tested?
Part 1:
- Copy `OCW_` env settings from RC
- Take a look at the `mit_learn_topics` field (3 distinct topics) and `topics` field (11 distinct topics) at https://ocw.mit.edu/courses/17-523-ethnicity-and-race-in-world-politics-fall-2005/data.json
- Run the following: `./manage.py backpopulate_ocw_data --course-name 17-523-ethnicity-and-race-in-world-politics-fall-2005`
- Go to http://api.open.odl.local:8063/api/v1/learning_resources_search/?q=%22Ethnicity+and+Race%22&yearly_decay_percent=0.  Check that:
  -  The `topics` field has 3 topics in the usual format with names "Anthropology", "Social Sciences" , and "Political Science" matching the 3 distinct topics in the ocw data.json `mit_learn_topics` field
  - The `ocw_topics` field has 11 topic names as a list, matching the distinct topics in the ocw data.json `topics` field.

Part 2:
- Take a look at the `mit_learn_topics` field (null) and `topics` field (10 distinct topics) at https://ocw.mit.edu/courses/12-003-atmosphere-ocean-and-climate-dynamics-fall-2008/data.json
- Run `./manage.py backpopulate_ocw_data --course-name 12-003-atmosphere-ocean-and-climate-dynamics-fall-2008`
- Go to http://api.open.odl.local:8063/api/v1/learning_resources_search/?q=%22Atmosphere,%20Ocean%20and%20Climate%20Dynamics%22&yearly_decay_percent=0 and check that:
  - The `topics` field is identical to the one here: https://api.learn.mit.edu/api/v1/courses/4840/
  - The `ocw_topics` field has the same 10 distinct topics as the data.json file

Part 3:
- Recreate the search index: `./manage.py recreate_index --all`
- Search for an ocw_topic, you should get the above result back: http://api.open.odl.local:8063/api/v1/learning_resources_search/?ocw_topic=Climate%20Studies

### Additional Context
- Will need to run a full `backpopulate_ocw_data` on RC and prod (with `--skip-contentfiles`), plus `rrecreate_index`
- Aggregations by ocw_type will be added in a subsequent PR (issue https://github.com/mitodl/hq/issues/5591) - not adding it yet because it would break search until reindexing is completed.
